### PR TITLE
Security fix for lack of rate limiting in hackathon-starter

### PR DESCRIPTION
**Pull request by arjunshibu**
### 📊 Metadata *

`hackathon-starter` is a boilerplate for Node.js web applications. It lacks rate-limiting, which allows an attacker to brute-force login credentials and mass account creation

### ⚙️ Description *

Lack of rate-limiting in the login and signup page of `hackathon-starter`

### 💻 Technical Description *

This fix uses the `express-rate-limit` package, which is a rate-limiting middleware for express based servers. I've applied it to POST requests issuing to `/login` and `/signup` routes.

#### Login Brute-Force

1. Install the package by following these [instructions](https://github.com/sahat/hackathon-starter#getting-started) or try the live demo https://hackathon-starter.walcony.com
2. Navigate to `/login` and send the POST request to Burp Intruder
3. Use Grep Match option and add string `Found. Redirecting to <a href="/login">/login</a>`
4. The succeeded payload will redirect to `/` instead of `/login`                                                
  https://raw.githubusercontent.com/arjunshibu/files/main/hackathon-starter-poc/poc1.png                                            
  https://raw.githubusercontent.com/arjunshibu/files/main/hackathon-starter-poc/poc2.png                                            
  https://raw.githubusercontent.com/arjunshibu/files/main/hackathon-starter-poc/poc3.png                                            
  
#### Mass Account Creation

1. Navigate to `/signup` and send the POST request to Burp Intruder
2. I used numbers as payloads
3. Use Grep Match option and add the string `Found. Redirecting to <a href="/">/</a>`
4. All succeeded payloads will redirect to `/` and to `/signup` if there is already a user                                                      
  https://raw.githubusercontent.com/arjunshibu/files/main/hackathon-starter-poc/poc4.png                                                       
  https://raw.githubusercontent.com/arjunshibu/files/main/hackathon-starter-poc/poc5.png

### 🔥 Proof of Fix (PoF) *

After fix any brute-forcing of login credentials and mass account creation is prevented by responding with `429 - Too Many Requests`
pof1
pof2

### 👍 User Acceptance Testing (UAT)

* I've executed unit tests.
* After fix the functionality is unaffected.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -44,6 +45,20 @@ const passportConfig = require('./config/passport');
...
+ * rate-limiters for /login and /signup POST requests
+ */
+const loginLimiter = rateLimit({
+  // block IP for 15 minutes after 20 requests
+  windowMs: 15 * 60 * 1000,
+  max: 20
+});
+const signupLimiter = rateLimit({
+  // block IP for 1 hour after 5 requests
+  windowMs: 60 * 60 * 1000,
```
Comment (Mik317) [inline app.js:58]: 1 hour is too much in my opinion ... we need to remember some users could have multiple problems during this phase (invalid pass, mail already registered etc ...)

**Review by Mik317 — APPROVED** [review]
LGTM :smile: :cake:
Screenshot from 2020-12-28 17-30-31

Please just change the period of time which blocks the `user` who's trying to rate limit :smile:
Also, @mufeedvh @bbeale @mzfr @toufik-airane what do you think about `times` ?

Cheers,
Mik

File: app.js (hunk 1 above)
Comment (arjunshibu) [reply app.js:58]: @Mik317 what about 1hour after 15requests?

Comment (mzfr) [reply app.js:58]: IMHO 15 requests every 15 min might be a good thing to keep. But to be 100% sure I think we should ask the project dev about this time limit.

**Comment by huntr-helper** [comment]
Congratulations arjunshibu - your fix has been selected! 🎉

Thanks for being part of the community & helping secure the world's open source code.
If you have any questions, please respond in the comments section, or hit us up on Discord. Your bounty is on its way - keep hunting!

[Come join us on Discord](https://huntr.dev/discord)

**Pull request merged**